pkg/token/execCredentialWriter.go

```diff
@@ -20,20 +39,67 @@ type execCredentialWriter struct{}
 
 // Write writes the ExecCredential to standard output for kubectl.
 func (*execCredentialWriter) Write(token adal.Token) error {
-	t := v1.NewTime(token.Expires())
-	ec := &v1beta1.ExecCredential{
-		TypeMeta: v1.TypeMeta{
-			APIVersion: "client.authentication.k8s.io/v1beta1",
-			Kind:       "ExecCredential",
-		},
-		Status: &v1beta1.ExecCredentialStatus{
-			Token:               token.AccessToken,
-			ExpirationTimestamp: &t,
-		},
+	fmt.Fprintln(os.Stderr, os.Getenv(execInfoEnv))
+	//os.Setenv(execInfoEnv, "TEST")
+	//fmt.Println("KUBERNETES_EXEC_INFO:", os.Getenv(execInfoEnv))
+	apiVersionFromEnv, err := helperGetApiVersionFromEnv()
+	if err != nil {
+		return err
 	}
+
+	var ec interface{}
+	t := metav1.NewTime(token.Expires())
+	if apiVersionFromEnv == "client.authentication.k8s.io/v1beta1" {
+		ec = &v1beta1.ExecCredential{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "client.authentication.k8s.io/v1beta1",
+				Kind:       "ExecCredential",
+			},
+			Status: &v1beta1.ExecCredentialStatus{
+				Token:               token.AccessToken,
+				ExpirationTimestamp: &t,
+			},
+		}
+	} else {
+		ec = &v1.ExecCredential{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "client.authentication.k8s.io/v1",
+				Kind:       "ExecCredential",
+			},
+			Status: &v1.ExecCredentialStatus{
+				Token:               token.AccessToken,
+				ExpirationTimestamp: &t,
+			},
+		}
+	}
+
 	e := json.NewEncoder(os.Stdout)
 	if err := e.Encode(ec); err != nil {
 		return fmt.Errorf("could not write the ExecCredential: %s", err)
 	}
 	return nil
 }
+
+func helperGetApiVersionFromEnv() (string, error) {
+	env := os.Getenv(execInfoEnv)
+	if env == "" {
+		return "client.authentication.k8s.io/v1beta1", nil
+	} else {
+		obj, _, err := codecs.UniversalDeserializer().Decode([]byte(env), nil, nil)
+		if err != nil {
+			return "", fmt.Errorf("decode: %w", err)
+		}
+		var execCredential clientauthentication.ExecCredential
+		if err := scheme.Convert(obj, &execCredential, nil); err != nil {
+			return "", fmt.Errorf("cannot convert to ExecCredential: %w", err)
+		}
+		if execCredential.TypeMeta.APIVersion == "" {
+			return "client.authentication.k8s.io/v1beta1", nil
+		}
+		if execCredential.TypeMeta.APIVersion == "client.authentication.k8s.io/v1beta1" || execCredential.TypeMeta.APIVersion == "client.authentication.k8s.io/v1" {
+			return execCredential.TypeMeta.APIVersion, nil
+		} else {
+			return "", errors.New("This api version is not supported")
```

Comment:
error message should indicate the provided api version

Reply:
Done already in current commit. Thanks!